### ISSUE TITLE
Make pub run faster

### ISSUE THREAD

**Issue by DartBot** [dart-lang/pub#1121]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="96" height="96"hspace="10"></img></a> **Issue by [nex3](https://github.com/nex3)**
_Originally opened as dart-lang/sdk#20859_

----

&quot;pub run&quot; is currently very slow when running local executables. We should do what we can to mitigate this, including:

\* Precompiling immutable transformers to snapshots on &quot;pub get&quot;
\* Precompiling immutable dependencies to source
\* Lazily load assets from untransformed dependencies
\* Don't load transformers at all if we can prove that they won't be used
\* Stop rewriting package imports (blocked on issue dart-lang/sdk#12474)
\* Load files directly rather than over HTTP (blocked on issue dart-lang/sdk#16368)

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Marked this as blocking dart-lang/sdk#20866.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

*Marked this as being blocked by dart-lang/sdk#20866.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

*Unmarked this as blocking dart-lang/sdk#20866.*

**Comment by DartBot** [comment]
<a href="https://github.com/kasperl"><img src="https://avatars.githubusercontent.com/u/2156198?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [kasperl](https://github.com/kasperl)**

----

*Added this to the [1.7](../milestones/1.7) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

r40392 caches immutable transformers, although they're not yet precompiled on "pub get".

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Issue dart-lang/pub#1063 has been merged into this issue.
______
 cc @iposva-google.
 cc @turnidge.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Removing Milestone-1.7. There's more performance improvements left to do, but "pub run" is about as fast as I was hoping for 1.7.
______
*Removed this from the [1.7](../milestones/1.7) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Also removing issue dart-lang/sdk#12474 as blocking, since it's now supported on the VM and that's all we need in pub.
______
*Unmarked this as being blocked by dart-lang/sdk#12474.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

With dart-lang/sdk@bf84233ad9a804020b56db93b305bce510b658ff making the yaml package faster, "pub run" is now acceptably fast. Fixing issue dart-lang/dart_style#301 and/or 20866 would still improve the speed substantially, though.
______
*Added [Fixed](../labels/Fixed) label.*
